Objective-C/CBLDatabase.mm

```diff
@@ -832,72 +832,70 @@ - (BOOL) saveDocument: (CBLDocument*)document
                 CBLStringBytes bDocID(document.id);
-                curDoc = c4doc_get(_c4db, bDocID, true, nil);
-                assert(curDoc);
+                C4Error err;
+                curDoc = c4doc_get(_c4db, bDocID, true, &err);
+                if (!curDoc) {
+                    if (err.code == kC4ErrorNotFound)
+                        return YES;
+                    return convertError(err, outError);
+                }
             }
             
-            // If deletion and the current doc has already been deleted:
-            if (deletion) {
-                if ((curDoc->flags & kDocDeleted) != 0) {
-                    [document replaceC4Doc: [CBLC4Document document: curDoc]];
-                    return YES;
+            if (![self saveDocument: document into: &newDoc
+                   withBaseDocument: nil asDeletion: deletion error: outError])
+                return NO;
+            
+            if (!newDoc) {
+                // Handle conflict:
+                if (concurrencyControl == kCBLConcurrencyControlOptimistic)
+                    return createError(CBLErrorConflict, outError);
+                
+                if (!curDoc) {
+                    CBLStringBytes bDocID(document.id);
+                    C4Error err;
+                    curDoc = c4doc_get(_c4db, bDocID, true, &err);
+                    assert(curDoc);
+                }
+                
+                // If deletion and the current doc has already been deleted:
+                if (deletion) {
+                   if ((curDoc->flags & kDocDeleted) != 0) {
+                        [document replaceC4Doc: [CBLC4Document document: curDoc]];
+                        curDoc = nil;
+                        return YES;
+                    }
                 }
+                
+                // Save changes on the current branch:
+                if (![self saveDocument: document into: &newDoc
+                       withBaseDocument: curDoc asDeletion: deletion error: outError])
+                    return NO;
             }
             
-            // Save changes on the current branch:
-            BOOL success = [self saveDocument: document into: &newDoc
-                             withBaseDocument: curDoc asDeletion: deletion error: outError];
-            c4doc_free(curDoc);
-            curDoc = nil;
-            if (!success)
-                return NO;
+            if (!transaction.commit())
+                return convertError(transaction.error(), outError);
+            
+            [document replaceC4Doc: [CBLC4Document document: newDoc]];
+            newDoc = nil;
+            return YES;
         }
-        
-        if (curDoc)
-            c4doc_free(curDoc);
-        
-        if (!transaction.commit()) {
-            c4doc_free(newDoc);
-            return convertError(transaction.error(), outError);
+        @finally {
+            if (curDoc)
```

Comment:
FYI, these `if` tests aren't necessary; all the `c4*free` and `c4*close` functions allow null arguments.